mempool/nop_mempool.go

```diff
@@ -0,0 +1,113 @@
+package mempool
+
+import (
+	"errors"
+
+	abcicli "github.com/cometbft/cometbft/abci/client"
+	abci "github.com/cometbft/cometbft/abci/types"
+	"github.com/cometbft/cometbft/internal/service"
+	"github.com/cometbft/cometbft/p2p"
+	"github.com/cometbft/cometbft/types"
+)
+
+// NopMempool is a mempool that does nothing.
+//
+// The ABCI app is responsible for storing, disseminating, and proposing transactions.
+// See [ADR-111](../docs/architecture/adr-111-nop-mempool.md).
+type NopMempool struct{}
+
+// ErrNotAllowed indicates that the operation is not allowed with `nop` mempool.
+var ErrNotAllowed = errors.New("not allowed with `nop` mempool")
+
+var _ Mempool = &NopMempool{}
+
+// CheckTx always returns ErrNotAllowed.
+func (*NopMempool) CheckTx(types.Tx) (*abcicli.ReqRes, error) {
+	return nil, ErrNotAllowed
+}
+
+// RemoveTxByKey always returns ErrNotAllowed.
+func (*NopMempool) RemoveTxByKey(types.TxKey) error { return ErrNotAllowed }
+
+// ReapMaxBytesMaxGas always returns nil.
+func (*NopMempool) ReapMaxBytesMaxGas(int64, int64) types.Txs { return nil }
+
+// ReapMaxTxs always returns nil.
+func (*NopMempool) ReapMaxTxs(int) types.Txs { return nil }
+
+// Lock does nothing.
+func (*NopMempool) Lock() {}
+
+// Unlock does nothing.
+func (*NopMempool) Unlock() {}
+
+// Update does nothing.
+func (*NopMempool) Update(
+	int64,
+	types.Txs,
+	[]*abci.ExecTxResult,
+	PreCheckFunc,
+	PostCheckFunc,
+) error {
+	return nil
+}
+
+// FlushAppConn does nothing.
+func (*NopMempool) FlushAppConn() error { return nil }
+
+// Flush does nothing.
+func (*NopMempool) Flush() {}
+
+// TxsAvailable always returns a closed channel.
+func (*NopMempool) TxsAvailable() <-chan struct{} {
+	txsAvailable := make(chan struct{}, 1)
```

Comment:
Minor. How often does consensus call this? Would it be worth creating (and closing) a channel upon construction and reusing it?

Reply:
Good point 👍  It's called quite often actually. Changed it to return `nil` https://github.com/cometbft/cometbft/pull/1643/commits/4044579adce425a720569403bf3c713310dd90fa